### ISSUE TITLE
NPE in ParserBolt

### ISSUE THREAD

**Issue by jakekdodd** [DigitalPebble/storm-crawler#48]
Encountered this while working on the Spiderling.

The NPE stems from a null content byte array, and occurs [here](https://github.com/DigitalPebble/storm-crawler/blob/1ad1cf5b526ac0d81df1e6e10dbad0b5c7492f61/src/main/java/com/digitalpebble/storm/crawler/bolt/ParserBolt.java#L168). I discovered that the offending URL was one that resulted in a 301 redirect. However, the protocol response's content is non-null for several other URLs that result in 301s, so the redirect response may not be the source of the problem.

Console output with stack trace:

```
52998 [Thread-12-parse] ERROR backtype.storm.util - Async loop died!
java.lang.RuntimeException: java.lang.NullPointerException
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:128) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.executor$fn__3441$fn__3453$fn__3500.invoke(executor.clj:748) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.util$async_loop$fn__464.invoke(util.clj:463) ~[storm-core-0.9.3.jar:0.9.3]
	at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
Caused by: java.lang.NullPointerException: null
	at com.digitalpebble.storm.crawler.bolt.ParserBolt.execute(ParserBolt.java:168) ~[storm-crawler-0.4-SNAPSHOT.jar:na]
	at backtype.storm.daemon.executor$fn__3441$tuple_action_fn__3443.invoke(executor.clj:633) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.executor$mk_task_receiver$fn__3364.invoke(executor.clj:401) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$clojure_handler$reify__1447.onEvent(disruptor.clj:58) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125) ~[storm-core-0.9.3.jar:0.9.3]
	... 6 common frames omitted
```
@jnioche have you seen this before?

**Comment by jnioche** [comment]
Never got that before. Do you know what URL caused it? Would be good to check that the Fetcher behaves as expected. 
This would not be an issue after committing #44 since the parser would be called only if the http code is Status.FETCHED. Having said that we should check that the content is not null and if so handle it via the status stream (see #43)

**Comment by jakekdodd** [comment]
Yep, the offending URL is http://www.dailynewslosangeles.com/.

I'm curious about why this particular case produces null content, when typically a 301 results in an empty byte array (I think). It's possible this is a bug in the HTTP implementation--but there might also be a meaningful difference between null content and empty content that we should report to Status.

**Comment by jnioche** [comment]
It doesn't happen now that I committed  #44 into trunk, however HttpResponse contains a null value for content indeed. 

This is caused by a SocketTimeoutException being thrown by readPlainContent in [HttpResponse](https://github.com/DigitalPebble/storm-crawler/blob/1ad1cf5b526ac0d81df1e6e10dbad0b5c7492f61/src/main/java/com/digitalpebble/storm/crawler/protocol/http/HttpResponse.java#L289). 
As you can see [here](https://github.com/DigitalPebble/storm-crawler/blob/1ad1cf5b526ac0d81df1e6e10dbad0b5c7492f61/src/main/java/com/digitalpebble/storm/crawler/protocol/http/HttpResponse.java#L230) we do nothing about this exception.
We should probably log this exception but also add an empty content array in the finally section so that we never get that null content.

The main issue however is in [readPlainContent](https://github.com/DigitalPebble/storm-crawler/blob/1ad1cf5b526ac0d81df1e6e10dbad0b5c7492f61/src/main/java/com/digitalpebble/storm/crawler/protocol/http/HttpResponse.java#L313) as we shouldn't even try to read from the stream at all if the contentLength is set to 0 by the server.

**Comment by jakekdodd** [comment]
Got it, good catch. I'll fix this

**Comment by jnioche** [comment]
Fixed in [master 4f1339e] NPE in ParserBolt #48
